GH-10090: Add channel adapters for AMQP 1.0

artembilan opened the pull request:
Related to: https://github.com/spring-projects/spring-integration/issues/10090

Spring AMQP now provides the `org.springframework.amqp:spring-rabbitmq-client` library to communicate to RabbitMQ with AMQP 1.0 protocol.

* Replace `spring-rabbit` with the `spring-rabbitmq-client` since the latter brings the AMQP 0.9 protocol library as transitive dependency
* Introduce an `AmqpClientMessageHandler` based on the `AsyncAmqpTemplate` where its implementation should be for AMQP 1.0, e.g. `org.springframework.amqp.rabbitmq.client.RabbitAmqpTemplate`
* This outbound channel adapter can behave as a gateway when `setRequiresReply(true)`
* Cover basic `AmqpClientMessageHandler` with integration tests
* Document this new feature

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

artembilan committed:
GH-10090: Add AmqpClientMessageHandler for AMQP 1.0

Related to: https://github.com/spring-projects/spring-integration/issues/10090

Spring AMQP now provides the `org.springframework.amqp:spring-rabbitmq-client` library
to communicate to RabbitMQ with AMQP 1.0 protocol.

* Replace `spring-rabbit` with the `spring-rabbitmq-client` since the latter brings
the AMQP 0.9 protocol library as transitive dependency
* Introduce an `AmqpClientMessageHandler` based on the `AsyncAmqpTemplate`
where its implementation should be for AMQP 1.0, e.g. `org.springframework.amqp.rabbitmq.client.RabbitAmqpTemplate`
* This outbound channel adapter can behave as a gateway when `setRequiresReply(true)`
* Cover basic `AmqpClientMessageHandler` with integration tests
* Document this new feature

cppwfs reviewed (changes requested):
Looks great!   
My usual nitpicks :-)

artembilan commented on spring-integration-amqp/src/main/java/org/springframework/integration/amqp/outbound/AmqpClientMessageHandler.java:
```diff
@@ -0,0 +1,317 @@
+		setRoutingKeyExpression(EXPRESSION_PARSER.parseExpression(routingKeyExpression));
+	}
+
+	public void setRoutingKeyExpression(Expression routingKeyExpression) {
```
No, if you chose to call these methods, then no null allowed.

cppwfs replied:
Understood.   Thank you!

artembilan commented on spring-integration-amqp/src/main/java/org/springframework/integration/amqp/outbound/AmqpClientMessageHandler.java:
```diff
@@ -0,0 +1,317 @@
+		}
+		this.evaluationContext = ExpressionUtils.createStandardEvaluationContext(getBeanFactory());
+
+		Assert.state(this.exchangeExpression == null || this.queueExpression == null,
```
Not sure what you mean.
The `this.exchangeExpression` is an `Expression`, not `String`

cppwfs replied:
DOH!    I was thinking of a string .  Apologies.

cppwfs commented on spring-integration-amqp/src/main/java/org/springframework/integration/amqp/outbound/AmqpClientMessageHandler.java:
```diff
@@ -0,0 +1,317 @@
+	}
+
+	@Override
+	protected void doInit() {
```
No worries at all...   
Just was thinking through my fingers :-)

cppwfs commented on spring-integration-amqp/src/test/java/org/springframework/integration/amqp/outbound/AmqpClientMessageHandlerTests.java:
```diff
@@ -0,0 +1,217 @@
+	AmqpClientMessageHandler amqpClientGateway;
+
+	@Test
+	void noExchangeOrQueue() {
```
I should say...  `neitherExchangeAndQueue`

artembilan commented on src/reference/antora/modules/ROOT/pages/amqp/amqp-1.0.adoc:
```diff
@@ -0,0 +1,55 @@
+Starting with version 7.0, Spring Integration provides channel adapters for RabbitMQ AMQP 1.0 support.
+These channel adapters are based on the `org.springframework.amqp:spring-rabbitmq-client` library.
+
+The Spring AMQP documentation talks about https://docs.spring.io/spring-amqp/reference/4.0/rabbitmq-amqp-client.html[RabbitMQ AMQP 1.0 support] in more detail.
```
Right. I'm not the best guy to judge English sentences, but what I know that technical, official documentation must be impersonated.
Therefore saying something like a command `see` is not good.
I know that we have a plethora of those `see`, `you`, `they`, `we` etc. throughout the doc, but that does not mean that we should introduce more.
WDYT?

cppwfs replied:
Sounds good.   We can hold off this change.

artembilan committed:
* Add `AmqpClientMessageProducer` implementation to consume message from RabbitMQ AMQP 1.0
* Fix typos and language in docs and Javadocs
* Add tests and docs about new `AmqpClientMessageProducer`
* Add `whats-new.adoc` bullet for a new `spring-integration-amqp` feature

cppwfs commented on spring-integration-amqp/src/test/java/org/springframework/integration/amqp/inbound/AmqpClientMessageProducerTests.java:
```diff
@@ -0,0 +1,180 @@
+	}
+
+	@Test
+	void receiveBatch() {
```
Do we want to have a receiveBatchAndAck test as well?

artembilan replied:
Sure!
Will look into that.
I think more about errors.
But let me fix `main` first!

cppwfs replied:
Sweet!   Thanks!

cppwfs reviewed (approved):
Changes to previous commit look great!
Addition of the AmqpClientMessageProducer look good as well.  Just one nitpick for an additional test.   

This some cool stuff!   Thanks for working on it!

artembilan committed:
* Fix `AmqpClientMessageProducer` for throwing a `ListenerExecutionFailedException`
when exception is bubbled from the downstream flow
* Add test for error handling
* Add test for batch manual ack
* Remove redundant (and dangerous) `@ComponentScan` from the `ManualAckTests`:
it does not do anything for the test suite, but is able to see all the `@Configuration`
classes in other tests of this package.
That may lead to unexpected behavior and failures

artembilan commented:
@cppwfs ,

let me know if you are going to merge this after your next review, or is that OK for me to continue piling related features into this PR.
Thanks

cppwfs reviewed (approved):
LGTM 
This is great stuff!
Thanks for adding the batch test!
